packages/studio-base/src/players/messageMemoryEstimation.ts

```diff
@@ -0,0 +1,148 @@
+  if (knownSize != undefined) {
+    return knownSize;
+  }
+
+  if (datatypes.size === 0) {
+    return OBJECT_BASE_SIZE; // Empty schema -> Empty object.
+  }
+
+  const definition = datatypes.get(typeName);
+  if (!definition) {
+    throw new Error(`Type '${typeName}' not found in definitions`);
+  }
+
+  let sizeInBytes = OBJECT_BASE_SIZE;
+
+  const nonConstantFields = definition.definitions.filter((field) => !(field.isConstant ?? false));
+  if (nonConstantFields.length > MAX_NUM_FAST_PROPERTIES) {
+    // If there are too many properties, V8 stores Objects in dictionary mode (slow properties)
+    // with each object having a self-contained dictionary. This dictionary contains the key, value
+    // and details of properties. Below we estimate the size of this additional dictionary. Formula
+    // adapted from
+    // medium.com/@bpmxmqd/v8-engine-jsobject-structure-analysis-and-memory-optimization-ideas-be30cfcdcd16
+    const propertiesDictSize =
+      16 + 5 * 8 + 2 ** Math.ceil(Math.log2((nonConstantFields.length + 2) * 1.5)) * 3 * 4;
+    sizeInBytes += propertiesDictSize;
+
+    // In return, properties are no longer stored in the properties array
+    sizeInBytes -= 4 * nonConstantFields.length;
+  }
+
+  for (const field of nonConstantFields) {
+    if (field.isComplex ?? false) {
+      const count =
+        field.isArray === true
+          ? 0 // We are conservative and assume an empty array
+          : 1;
+
+      const knownFieldSize = knownTypeSizes.get(field.type);
+      if (knownFieldSize != undefined) {
+        sizeInBytes += count > 0 ? count * knownFieldSize : OBJECT_BASE_SIZE;
+        continue;
+      }
+
+      if (checkedTypes != undefined && checkedTypes.includes(field.type)) {
+        // E.g. protobuf allows types to reference itself.
+        // For that reason we bail out here to avoid an infinite loop.
+        continue;
+      }
+
+      const complexTypeObjectSize = estimateMessageObjectSize(
+        datatypes,
+        field.type,
+        knownTypeSizes,
+        (checkedTypes ?? []).concat(field.type),
+      );
+      sizeInBytes += count > 0 ? count * complexTypeObjectSize : OBJECT_BASE_SIZE;
+    } else if (field.isArray === true) {
+      const arrayLength = field.arrayLength ?? 0; // We are conservative and assume an empty array;
+      switch (field.type) {
+        // Assume that fields get deserialized as typed arrays
+        case "int8":
+        case "uint8":
+          sizeInBytes += TYPED_ARRAY_BASE_SIZE + arrayLength * 1;
+          break;
+        case "int16":
+        case "uint16":
+          sizeInBytes += TYPED_ARRAY_BASE_SIZE + arrayLength * 2;
+          break;
+        case "int32":
+        case "uint32":
+        case "float32":
+          sizeInBytes += TYPED_ARRAY_BASE_SIZE + arrayLength * 4;
+          break;
+        case "float64":
+        case "int64":
+        case "uint64":
+          sizeInBytes += TYPED_ARRAY_BASE_SIZE + arrayLength * 8;
```

Comment:
nit: this could be reduced a bit
```suggestion
        case "int8":
        case "uint8":
        case "int16":
        case "uint16":
        case "int32":
        case "uint32":
        case "float32":
        case "float64":
        case "int64":
        case "uint64":
          sizeInBytes += TYPED_ARRAY_BASE_SIZE + arrayLength * FIELD_SIZE_BY_PRIMITIVE[field.type];
          break;
```
or have a set of typedarray types and check to see if the field.type is in it

Reply:
oop the field size by primitive probably doesn't apply here.